### PR TITLE
os/bluestore: osd scrape read/write operations

### PR DESCRIPTION
**Pull request by pereman2**
To run:
```bash
../src/test/objectstore/allocsim/ops_scraper.py --ceph_bin_path bin/ceph
```

Example:
```bash
$ ../src/test/objectstore/allocsim/ops_scraper.py --ceph_bin_path bin/ceph                                                                           [13:06:39]
collecting from osd. 0
op_type | offset/extent | name | pgid | initiated_at | who
-------------------------------------------------------
write 0~4194304 4:c4b97d2c:::be151868021c187609fb97e4c22ae1f9:head 4.3 2023-07-12T13:06:34.959082+0200 client.4831.0:351
write 0~4194304 4:d392075e:::b068dce954dca8ca56d72e02e87f39c8:head 4.b 2023-07-12T13:06:35.093439+0200 client.4831.0:358
write 0~4194304 4:bc1ac643:::7b815bd25a40695c891dc06bdf2b65be:head 4.1d 2023-07-12T13:06:35.128097+0200 client.4831.0:360
write 0~4194304 4:d0d21bd5:::cc531199dd2707370bb461cce82bdef6:head 4.b 2023-07-12T13:06:36.023082+0200 client.4831.0:367
write 0~4194304 4:a9abc92c:::0703d6f43d14bc33d69f94ecb96c522c:head 4.15 2023-07-12T13:06:36.367886+0200 client.4831.0:369
write 0~4194304 4:9b4684ea:::d5856dc0ecad96f811012e0937d5e0ac:head 4.19 2023-07-12T13:06:36.654584+0200 client.4831.0:370
write 0~4194304 4:d15b8c87:::c3344e5de8c8936876965d9fb056ad89:head 4.b 2023-07-12T13:06:36.835690+0200 client.4831.0:371
write 0~4194304 4:ad41de2a:::8a1e2743b577a67d6bc8a1d514391b53:head 4.15 2023-07-12T13:06:37.102741+0200 client.4831.0:373
write 0~4194304 4:fc35d54f:::39cd961e7e33680f4bfd1ca0fe43dda4:head 4.1f 2023-07-12T13:06:38.047612+0200 client.4831.0:377

collecting from osd. 1
op_type | offset/extent | name | pgid | initiated_at | who
-------------------------------------------------------
write 0~4194304 4:4c9ed2e8:::86206ccc35d8aceaf78670a751a6b550:head 4.12 2023-07-12T13:06:34.771423+0200 client.4831.0:340
write 0~4194304 4:106d9524:::fe7534cc604b0eb50498cb23aaca9161:head 4.8 2023-07-12T13:06:35.128105+0200 client.4831.0:359
write 0~4194304 4:a6e3fdb1:::2692491bfc4c584ca5b1efcb896936c2:head 4.5 2023-07-12T13:06:35.567070+0200 client.4831.0:363
write 0~4194304 4:2d9a471a:::ba51255646d00c496c2638bad03e0869:head 4.14 2023-07-12T13:06:35.568290+0200 client.4831.0:364
write 0~4194304 4:2b1381b1:::c58fa31e5d3a4f5550ee143bcb07a49f:head 4.14 2023-07-12T13:06:35.764699+0200 client.4831.0:365
write 0~4194304 4:a22e158f:::b2116e08700da18595b96d58120fe81c:head 4.5 2023-07-12T13:06:36.023101+0200 client.4831.0:366
```

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by aclamk** [comment]
Time "2023-07-12T13:06:36.023101+0200" should be converted to seconds (but preserving 10^-6 precision).
Omit the timezone.

File: src/test/objectstore/allocsim/ops_scraper.py [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+            prog='OSD operations parser')
+    parser.add_argument('--nohash',
+                        action='store_true', required=False)
+    parser.add_argument('--debug_level', type=str, default='warning')
+    parser.add_argument('--ceph_bin_path', required=False)
+    parser.add_argument('--freq', required=False, type=int, default=1)
+    parser.add_argument('--min_history_size', required=False, type=int, default=100)
+    parser.add_argument('--max_history_size', required=False, type=int, default=1000)
+    parser.add_argument('--out', required=False, help="filename to write output to. If none is provided it will be written to stdout")
+    args = parser.parse_args()
```
Comment (aclamk) [inline src/test/objectstore/allocsim/ops_scraper.py:191]: We need to give an option to select osds to capture ops.
The tool should *not* start capturing all osds in default mode.

File: src/test/objectstore/allocsim/ops_scraper.py [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+    Env.setup_env(args)
+
+    log_levels = {
+            'CRITICAL': logging.CRITICAL,
+            'ERROR': logging.ERROR,
+            'WARNING': logging.WARNING,
+            'INFO': logging.INFO,
+            'DEBUG': logging.DEBUG,
+            'NOTSET': logging.NOTSET
+    }
```
Comment (aclamk) [inline src/test/objectstore/allocsim/ops_scraper.py:202]: Names for debug levels are good, but in output.
I would imagine that one would prefer to do "-v -v -v -v" or "--debug=4" rather then check the naming "--debug=info" and then changing "--debug=debug" when that turned to be not enough.

File: src/test/objectstore/allocsim/ops_scraper.py [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+
+    Env.setup_env(args)
+
+    log_levels = {
+            'CRITICAL': logging.CRITICAL,
+            'ERROR': logging.ERROR,
+            'WARNING': logging.WARNING,
+            'INFO': logging.INFO,
+            'DEBUG': logging.DEBUG,
+            'NOTSET': logging.NOTSET
```
Comment (aclamk) [inline src/test/objectstore/allocsim/ops_scraper.py:201]: NOTSET logic (as I imagine means print nothing) should be above CRITICAL, not below DEBUG.

Comment (pereman2) [reply src/test/objectstore/allocsim/ops_scraper.py]: it means all debug levels

File: src/test/objectstore/allocsim/ops_scraper.py [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+            logger.info(f'osd.{osd.name} command dump_historic_ops with {process_info.ops_count} ops took {process_info.command_time}')
+
+            osd.periods += 1
+            if (osd.periods % update_period) == 0:
+                ops_per_period = (osd.sum_ops / osd.periods)
+                new_history_size = int(ops_per_period * history_overlap)
+                if new_history_size >= min_history_size and new_history_size <= max_history_size:
+                    _set_osd_history_size(osd.name, new_history_size)
+                    logger.info(f'changing osd.{osd.name} history size from {history_size} to {new_history_size}')
+                    history_size = new_history_size
```
Comment (aclamk) [inline src/test/objectstore/allocsim/ops_scraper.py:286]: I imagined that both `capture_period` and `history_size` will be a values that are adapting per-OSD.
So the shape of the main loop will be something like:
```
while true:
  sleep(time until nearest OSD needs to be captured)
  for all OSDs:
    if osd is due:
      capture
      write captured to outfile
      adapt capture params if necessary (frequency, history_size)
      recalculate new sleep time for main loop
```
Can the "adapt capture params" step be moved to separate function?

File: src/test/objectstore/allocsim/ops_scraper.py [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+
+    osd_processing_start = time.time()
+
+    operations_str = 'op_type | offset/extent | name | pgid | initiated_at | who\n'
+    operations_str += '-------------------------------------------------------\n'
+    new_ops = 0 
+    oldest_operation = None
+    capture_period_start = _to_timestamp(Env.store()[osd_id]['last_op'])
+    for op in historic['ops']:
+        initiated_at = datetime.datetime.fromisoformat(op['initiated_at'])
```
Comment (aclamk) [inline src/test/objectstore/allocsim/ops_scraper.py:155]: Traceback (most recent call last):
  File "../src/test/objectstore/allocsim/ops_scraper.py", line 303, in <module>
    main()
  File "../src/test/objectstore/allocsim/ops_scraper.py", line 250, in main
    process_info = process_osd(osd.name)
  File "../src/test/objectstore/allocsim/ops_scraper.py", line 155, in process_osd
    initiated_at = datetime.datetime.fromisoformat(op['initiated_at'])
AttributeError: type object 'datetime.datetime' has no attribute 'fromisoformat'

File: src/test/objectstore/allocsim/ops_scraper.py [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+    logger.info(f'collecting from osd.{osd_id}')
+    command_time_start = time.time()
+    historic: Dict[str, Any] = run_ceph_command(['tell', f'osd.{osd_id}', 'dump_historic_ops'])
+    command_time = time.time() - command_time_start
+
+    osd_processing_start = time.time()
+
+    operations_str = 'op_type | offset/extent | name | pgid | initiated_at | who\n'
+    operations_str += '-------------------------------------------------------\n'
+    new_ops = 0
```
Comment (aclamk) [inline src/test/objectstore/allocsim/ops_scraper.py:151]: output helper does not match with content:
```
op_type | offset/extent | name | pgid | initiated_at | who
-------------------------------------------------------
2023-09-05 13:25:52.950495+00:00 client.4318.0:49128 write 0~4096 4:67411e1c:::test_object_589:head 4.6
2023-09-05 13:25:54.143097+00:00 client.4318.0:49216 write 0~4096 4:34367256:::test_object_677:head 4.c
2023-09-05 13:25:54.310721+00:00 client.4318.0:49231 write 0~4096 4:ba7431fb:::test_object_689:head 4.1d
```

**Pull request merged**